src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala

```diff
@@ -0,0 +1,619 @@
+    val key4: Array[Byte] = Hex.decode("123500")
+    val trie = EmptyTrie.put(key1, key1).put(key2, key2).put(key3, key3).put(key4, key4)
+    val wrongSource = HashMapDataSource().update(
+      rootHash = Array.emptyByteArray,
+      toRemove = Seq(),
+      toUpdate = Seq(trie.getRootHash -> trie.dataSource.get(trie.getRootHash).get)
+    )
+    val trieWithWrongSource = MerklePatriciaTrie[Array[Byte], Array[Byte]](trie.getRootHash, wrongSource, hashFn)
+    val trieAfterDelete = Try {
+      trieWithWrongSource.remove(key1)
+    }
+    assert(trieAfterDelete.isFailure)
+  }
+
+  test("Get in an empty trie"){
+    val obtained = EmptyTrie.get(Hex.decode("1234"))
+    assert(obtained.isEmpty)
+  }
+
+  test("Insert with an empty key in a branch node"){
+    val key1: Array[Byte] = Hex.decode("00")
+    val key2: Array[Byte] = Hex.decode("f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val key3: Array[Byte] = Hex.decode("")
+    val val3: Array[Byte] = Hex.decode("1234")
+    val trie = trieWithBranch.put(key3, val3)
+    val key1Get = trie.get(key1)
+    val key2Get = trie.get(key2)
+    val key3Get = trie.get(key3)
+    assert(key1Get.isDefined && (key1Get.get sameElements key1))
+    assert(key2Get.isDefined && (key2Get.get sameElements key2))
+    assert(key3Get.isDefined && (key3Get.get sameElements val3))
+  }
+
+  test("Remove of a key (not in trie) whose value should be in a branch node"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  test("Remove of a key (not in trie) that should be in the child of a BranchNode that is not present"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11a0"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  /* IODB tests */
+  test("Simple test with IODB as Source") {
+    //create temporary dir
+    val dir = File.createTempFile("iodb", "iodb")
+    dir.delete()
+    dir.mkdir()
+
+    //open new store
+    val dataSource = new IodbDataSource(new LSMStore(dir = dir, keySize = 32))
+    val emptyTrie = MerklePatriciaTrie[Int, Int](dataSource, hashFn)
+    val trieWithOneElement = emptyTrie.put(1, 5)
+    val obtained = trieWithOneElement.get(1)
+    assert(obtained.isDefined)
+    assert(obtained.get == 5)
+    val trieAfterDelete = trieWithOneElement.remove(1)
+    val obtainedAfterDelete = trieAfterDelete.get(1)
+    assert(obtainedAfterDelete.isEmpty)
+  }
+
+  ignore("IODB test - Insert of the first 5000 numbers hashed and then remove half of them"){
```

Comment:
> 90 sec does not looks scary to me ;)

It may be scary once we accumulate more of such tests. I wouldn't want to wait over an hour for a PR to go green.

> in order to finish this PR CircleCI will have to successfully run an additional step - sbt it:test

@whysoserious it looks that we need a proper `circle.yml` file first (I thought Circle was smart enough to pick up `travis.yml` but apparently it's just using some defaults for Scala). So I'd say that's a follow-up PR.

Anyway if those tests currently take just a few minutes then it's OK to require them for every PR, but _ultimately_ I'd like that we run them conditionally: either when merging to master or during nightly builds.